src/cores/account/Account.sol

```diff
@@ -0,0 +1,216 @@
+            // return if signature is malformed
+            if (err != ECDSA.RecoverError.NoError) return 1;
+
+            // authenticate signer, terminating early with status code 1 on failure
+            bool validSigner = _verifySigner(signer);
+            if (!validSigner) return 1;
+        }
+
+        /// @notice BLS sig aggregator and timestamp expiry are not currently supported by this contract 
+        /// so `bytes20(0x0)` and `bytes6(0x0)` suffice. To enable support for aggregator and timestamp expiry,
+        /// override the following params
+        bytes20 authorizer;
+        bytes6 validUntil;
+        bytes6 validAfter;
+
+        validationData = uint256(bytes32(abi.encodePacked(authorizer, validUntil, validAfter)));
+
+        /// @notice nonce collision is managed entirely by the EntryPoint, but validation hook optionality
+        /// for child contracts is provided here as `_checkNonce()` may be overridden
+        _checkNonce();
+
+        // check fee payment
+        if (missingAccountFunds != 0) {
+            _preFund(missingAccountFunds);
+        }
+    }
+
+    /// @dev Function to recover a signer address from the provided hash and signature
+    /// and then verify whether the recovered signer address is a recognized Turnkey 
+    /// @param hash The 32 byte digest derived by hashing signed message data. Sadly, name is canonical in ERC1271.
+    /// @param signature The signature to be verified via recovery. Must be prepended with validator address
+    function isValidSignature(bytes32 hash, bytes memory signature) public view returns (bytes4 magicValue) {
+        // try extracting packed validator data to check for modular validation format
+        bytes32 data;
+        assembly { data := mload(add(signature, 0x20)) }
+        (bytes8 flag, address validator) = (bytes8(data), address(uint160(uint256(data))));
+
+        // collision of a signature's first 8 bytes with flag is very unlikely; impossible when incl validator address
+        if (flag == VALIDATOR_FLAG && isValidator(validator)) {
+            ( , bytes memory nestedSig) = abi.decode(signature, (bytes32, bytes));
+
+            // format call for Validator module
+            bytes4 ret = IValidator(validator).isValidSignature(hash, nestedSig);
+
+            // validator will return either correct `magicValue` or error code `INVALID_SIGNER`
+            magicValue = ret;
+        } else {
+            // support non-modular signatures by recovering signer address and reverting malleable or invalid signatures
+            (address signer, ECDSA.RecoverError err) = ECDSA.tryRecover(hash, signature);
+            // return if signature is malformed
+            if (err != ECDSA.RecoverError.NoError) return bytes4(0);
+
+            // authenticate signer
+            bool validSigner = _verifySigner(signer);
+            magicValue = validSigner ? this.isValidSignature.selector : bytes4(hex'ffffffff');
+        }
```

Comment:
As a less opinionated abstraction, can we make the majority of this else block into a function that returns a `isValid` bool? If we're already creating room for people to customize what a verified signer is with `_verifySigner`, we can give more control by giving that modular function more responsibility. How does this look?
```suggestion
        } else {
            // support custom default signature handling
            bool isValid = _defaultIsValidSignature(hash, signature);
            magicValue = isValid ? this.isValidSignature.selector : bytes4(hex'ffffffff');
        }
```

We replace a single `_verifySigner(address signer)` with two default functions for both validation patterns:

```
function _defaultIsValidSignature(bytes32 hash, bytes memory signature) internal returns (bool);
function _defaultValidateUserOp(UserOperation calldata userOp, bytes32 userOpHash, uint256 missingAccountFunds);
```

This opens up customization to change default behavior if our modular validator pattern is not being used, making the decision to enroll in the system itself more modular.

Reply:
Like this a lot, thanks for the suggestion! Resolved in 48c8d47